README.md
# Leaflet.GridLayer.GoogleMutant

A [LeafletJS](http://leafletjs.com/) plugin to use Google maps basemaps.

## Demo

[https://ivansanchez.gitlab.io/Leaflet.GridLayer.GoogleMutant/demo.html](https://ivansanchez.gitlab.io/Leaflet.GridLayer.GoogleMutant/demo.html)

## Compatibility

-   This plugin doesn't work on IE10 or lower, as [that browser doesn't implement DOM mutation observers](https://caniuse.com/#feat=mutationobserver). Chrome, Firefox, Safari, IE11 and Edge are fine.

-   Starting with v0.11.0, the code relies on `Symbol` and `Map`. IE11 and [browsers that don't support `Symbol`](https://www.caniuse.com/mdn-javascript_builtins_symbol) or [don't support `Map`](https://www.caniuse.com/mdn-javascript_builtins_map) also need polyfills to work.

-   The `maxNativeZoom` functionality introduced in v0.5.0 (thanks, @luiscamacho!) requires Leaflet >1.0.3.

## Usage

Include the GMaps JS API in your HTML, plus Leaflet:

```html
<script src="https://maps.googleapis.com/maps/api/js?key=YOUR_API_KEY" async defer></script>
<link
	rel="stylesheet"
	href="https://unpkg.com/leaflet@1.7.1/dist/leaflet.css"
	integrity="sha512-xodZBNTC5n17Xt2atTPuE1HxjVMSvLVW9ocqUKLsCC5CXdbqCmblAshOMAS6/keqq/sMZMZ19scR4PsZChSR7A=="
	crossorigin=""
/>
<script
	src="https://unpkg.com/leaflet@1.7.1/dist/leaflet-src.js"
	integrity="sha512-I5Hd7FcJ9rZkH7uD01G3AjsuzFy3gqz7HIJvzFZGFt2mrCS4Piw9bYZvCgUE0aiJuiZFYIJIwpbNnDIM6ohTrg=="
	crossorigin=""
></script>
```

Include the GoogleMutant javascript file:

```html
<script src="https://unpkg.com/leaflet.gridlayer.googlemutant@latest/dist/Leaflet.GoogleMutant.js"></script>
```

Then, you can create an instance of `L.GridLayer.GoogleMutant` on your JS code:

```javascript
var roads = L.gridLayer
	.googleMutant({
		type: "roadmap", // valid values are 'roadmap', 'satellite', 'terrain' and 'hybrid'
	})
	.addTo(map);
```

It's also possible to use [custom styling](https://developers.google.com/maps/documentation/javascript/styling)
by passing a value to the `styles` option, e.g.:

```javascript
var styled = L.gridLayer
	.googleMutant({
		type: "roadmap",
		styles: [
			{ elementType: "labels", stylers: [{ visibility: "off" }] },
			{ featureType: "water", stylers: [{ color: "#444444" }] },
		],
	})
	.addTo(map);
```

## Installing a local copy

If you don't want to rely on a CDN to load GoogleMutant, you can:

-   Fetch it with [NPM](https://www.npmjs.com/) by running `npm install --save leaflet.gridlayer.googlemutant`.
-   Fetch it with [Yarn](https://yarnpkg.com/) by running `yarn add leaflet.gridlayer.googlemutant`.
-   We discourage using [Bower](https://bower.io/) but, if you must, `bower install https://gitlab.com/IvanSanchez/Leaflet.GridLayer.GoogleMutant.git`.

You can also download a static copy from the CDN, or clone this git repo.

## Known caveats

-   `hybrid` mode prunes tiles before needed for no apparent reason, so the map flickers when there is a zoom change.

-   Even though imagery exists at zoom levels up to 23, GoogleMutant caps the max zoom level at 22.

    This is to prevent scenarios where detecting imagery at those zooms levels is hard and creates problems (e.g. when zooming in/out close to a the boundary of such hi-res imagery).

    You can override this (at your own risk!) by using the `maxZoom` option with a value larger than 22.

-   GoogleMutant is meant to provide a reliable (and ToC-compliant) way of loading Google Map's tiles into Leaflet, nothing more.

    This means that route finding, geocoding, POI info, streetview, KML support, and in general anything that depends on calls to the Google Maps API are **not** implemented and are **not** a priority.

## Motivation

Before GoogleMutant, it was already possible to display Google Maps in Leaflet, but unfortunately the state of the art was far from perfect:

-   [Shramov's Leaflet plugin implementation](https://github.com/shramov/leaflet-plugins) (as well as an old, not recommended [OpenLayers technique](http://openlayers.org/en/v3.0.0/examples/google-map.html)) suffer from a [big drawback](https://github.com/shramov/leaflet-plugins/issues/111): the basemap and whatever overlays are on top are _off sync_. This is very noticeable when dragging or zooming.
-   [MapGear's implementation with OpenLayers](https://github.com/mapgears/ol3-google-maps) uses a different technique (decorate OL3 with GMaps methods), but has a different set of [limitations](https://github.com/mapgears/ol3-google-maps/blob/master/LIMITATIONS.md).
-   [Avin Mathew's implementation](https://avinmathew.com/leaflet-and-google-maps/) uses a clever timer-based technique, but it requires jQuery and still feels jittery due to the timers.

Before, an instance of the Google Maps JS API was displayed behind the Leaflet container, and synchronized as best as it could be done.

Now, in order to provide the best Leaflet experience, GoogleMutant uses both [DOM mutation observers](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) and `L.GridLayer` from Leaflet 1.0.0. The basemap tiles are still requested _through_ the Google maps JavaScript API, but they switch places to use Leaflet drag and zoom.

## Legalese

---

"THE BEER-WARE LICENSE":
<ivan@sanchezortega.es> wrote this file. As long as you retain this notice you
can do whatever you want with this stuff. If we meet some day, and you think
this stuff is worth it, you can buy me a beer in return.

---


CHANGELOG

# v0.14.0

* Updated regexps to cover a rare case.

# v0.13.5 (2022-02-17)

* Changed default of `style` option from empty object to empty array to prevent
  crash with latest upstream.

# v0.13.4 (2021-05-05)

* Fix race condition in _initMutant when layer was removed before api was ready

# v0.13.3 (2021-04-25)

* _update: process `center` argument.
  (initially implemented in 11c23155 but was non-functional due to a bug)

# v0.13.2 (2021-04-24)

* Added a .npmignore file to shrink the size of the NPM package

# v0.13.0 (2021-04-24)

* Implemented `whenReady` method (by @johnd0e)
* Removed `Promise` dependency (by @johnd0e)
* Fixed a regression tiles not loading on fractional zoom (due to cleanup of _resize())

# v0.12.1

* Rebuilt; dist script should be up-to-date in CDNs.

# v0.12.0

* Wrap GAPIPromise into a Singleton pattern (by @louisnow), should fix race conditions when loading JS libraries in weird orders or using bundlers or so
* Load more offscreen tiles when fractional zoom is enabled (150%→180%) to prevent artifacts

# v0.11.3 (2020-12-18)

* Fixed visibility/clickability of attribution and logo

# v0.11.2 (2020-12-14)

* Fixed tile alignment when map is inside a 'text-align' block.

# v0.11.1 (2020-12-14)

* Fixed a greedy map invalidation condition when removing/adding mutants to a map.

# v0.11.0 (2020-12-14)

* Refactored a bunch of functionality with a LRU cache
* Switched from `eslint` to `prettier`.

# v0.10.2

* Fix race condition when pruning a tile when the mutant has been removed from the map.

# v0.10.1

* Tweaks for iOS14 compatibility (by @vonrussel)

# v0.10.0

* Fixed IE11 compatibility (by @utilmind)
* Code cleanups (by @johnd0e)

# v0.9.0

* Increase tolerance of tile pruning algorithm (by @Yassin)

# v0.8.0

* Reuse mutant instances to lower calls to google APIs (by @redneb)

# v0.7.1 (2018-12-19)

* Remove "Google maps can't load correctly" message (on bad API key)
* Update readme to display Leaflet 1.3.4.
* Fix setOpacity() calls
* Remove "For development purposes only" overlaid tiles on aerial/hybrid
* Honour updateWhenIdle option for GridLayers

# v0.7.0 (2018-06-06)

* Remove "Sorry, we have no imagery here" empty <div>s whenever they mutate into the DOM
* Refactor away one `.bind()` call on every mutated element

# v0.6.4 (2017-09-18)

* Fix a reference error when destroying the map under some frameworks (by @mkochendorfer)

# v0.6.3

* Clear listeners for `idle` events in the mutant (by @ryan-morris)
* Remove legacy `L.Events.Mixin` (by @luiscamacho)
* Change default `maxZoom` option (by @luiscamacho)

# v0.6.2

* Specify Leaflet as a peerDependency in package.json

# v0.6.1

* Fix missing tiles by keeping tiles more agressively (by @luiscamacho)

# v0.6.0

* Add support for displaying some types of the traffic and "pegman" layers (by @luiscamacho)
* Add more sanity checks, prevent errors when the GMaps code takes time to load (by @tedbreen and @zakjan)
* Fixed a `forEach` iterator causing trouble in IE11 (by @dnasir)
* Fixed some race conditions by copying the tiles instead of moving them (by @luiscamacho)

# v0.5.1

* Handle `maxNativeZoom` based on the available imagery (by @luiscamacho)

# v0.5.0

* Handle `maxNativeZoom` based on the available imagery (by @luiscamacho)

# v0.4.5

* Shift controls in the lower left corner (in addition to those in the lower right) (by @luiscamacho).
* Resolve tiles in the same order they are mutated in, fixes race conditions at zooms 0-1 (by @TolonUK).
* Explicitly set z-index of mutant to avoid event propagation problems on iOS (by @robertharm).

# v0.4.4

* Changed README to show how to polyfill promises for IE11 (by @dullroar).
* Fixed a race condition on `_freshTiles` handling (by @luiscamacho).

# v0.4.3

* Fixed displacement of label tiles in hybrid mode

# v0.4.2

* Fixed typos in variable references
* Added linting

# v0.4.1

* Fixed a race condition involving spurious tile caching (thanks @tyllo)

# v0.4.0

* Ability to used styled maps (by @burakcalik and @danvirsen)

# v0.3.1

* Minor bugfixes by @tomhughes

# v0.3.0

* Implemented 'hybrid' mode, though there are tile pruning artifacts

# v0.2.1

* Fix entry point in package.json

# v0.2.0

* Images from the static maps API (which only appear on small viewports) are now removed and thus do not obscure map controls.

# v0.1.0

* Initial release


src/Leaflet.GoogleMutant.js
// GoogleMutant by Iván Sánchez Ortega <ivan@sanchezortega.es>

// Based on https://github.com/shramov/leaflet-plugins
// GridLayer like https://avinmathew.com/leaflet-and-google-maps/ , but using MutationObserver instead of jQuery

/*
"THE BEER-WARE LICENSE":
<ivan@sanchezortega.es> wrote this file. As long as you retain this notice you
can do whatever you want with this stuff. If we meet some day, and you think
this stuff is worth it, you can buy me a beer in return.
*/

import { LRUMap } from "./lru_map.js";

// 🍂class GridLayer.GoogleMutant
// 🍂extends GridLayer
L.GridLayer.GoogleMutant = L.GridLayer.extend({
	options: {
		maxZoom: 21, // can be 23, but ugly if more than maxNativeZoom
		// 🍂option type: String = 'roadmap'
		// Google's map type. Valid values are 'roadmap', 'satellite', 'terrain' or 'hybrid'.
		type: "roadmap",
		maxNativeZoom: 21,
	},

	initialize: function (options) {
		L.GridLayer.prototype.initialize.call(this, options);

		// Couple data structures indexed by tile key
		this._tileCallbacks = {}; // Callbacks for promises for tiles that are expected
		this._lru = new LRUMap(100); // Tile LRU cache

		this._imagesPerTile = this.options.type === "hybrid" ? 2 : 1;

		this._boundOnMutatedImage = this._onMutatedImage.bind(this);
	},

	onAdd: function (map) {
		L.GridLayer.prototype.onAdd.call(this, map);
		if (this._mutantContainer) {
			this._mutantOnAdd();
		} else { // first run
			this._initMutantContainer();
			this.initApi(this._initMutant.bind(this));
		}
	},

	onRemove: function (map) {
		L.GridLayer.prototype.onRemove.call(this, map);
		map._container.removeChild(this._mutantContainer);
		if (!this._mutant) { return; }

		this._observer.disconnect();

		if (this._logoContainer) {
			L.DomUtil.remove(this._logoContainer);
		}
		if (this._attributionContainer) {
			L.DomUtil.remove(this._attributionContainer);
		}
		google.maps.event.clearListeners(this._mutant, "idle");
	},

	// 🍂method addGoogleLayer(name: String, options?: Object): this
	// Adds layer with the given name and options to the google Map instance.
	// `name`: one of the google maps API layers, with it's constructor available in `google.maps` object.
	// currently following values supported: 'TrafficLayer', 'TransitLayer', 'BicyclingLayer'.
	// `options`: see https://developers.google.com/maps/documentation/javascript/reference/map
	addGoogleLayer: function (googleLayerName, options) {
		if (!this._subLayers) this._subLayers = {};
		this.whenReady(() => {
			var Constructor = google.maps[googleLayerName];
			var googleLayer = new Constructor(options);
			googleLayer.setMap(this._mutant);
			this._subLayers[googleLayerName] = googleLayer;
		});
		return this;
	},

	// 🍂method removeGoogleLayer(name: String): this
	// Removes layer with the given name from the google Map instance.
	removeGoogleLayer: function (googleLayerName) {
		this.whenReady(() => {
			var googleLayer = this._subLayers && this._subLayers[googleLayerName];
			if (googleLayer) {
				googleLayer.setMap(null);
				delete this._subLayers[googleLayerName];
			}
		});
		return this;
	},

	_initMutantContainer: function () {
		this._mutantContainer = L.DomUtil.create(
			"div",
			"leaflet-google-mutant leaflet-top leaflet-left"
		);
		this._mutantContainer.id = "_MutantContainer_" + L.Util.stamp(this._mutantContainer);
		this._mutantContainer.style.pointerEvents = "none";

		L.DomEvent.off(this._mutantContainer);
	},

	_mutantOnAdd: function () {
		if (!this._mutant) { return; }
		this._map.getContainer().appendChild(this._mutantContainer);

		this.setOpacity(this.options.opacity);
		const style = this._mutantContainer.style;
		if (this._map.options.zoomSnap < 1) {
			// Fractional zoom needs a bigger mutant container in order to load more (smaller) tiles
			style.width = "180%";
			style.height = "180%";
		} else {
			style.width = "100%";
			style.height = "100%";
		}
		style.zIndex = -1;

		this._attachObserver(this._mutantContainer);

		// Attribution and logo nodes are not mutated a second time if the
		// mutant is removed and re-added to the map, hence they are
		// not cleaned up on layer removal, so they can be added here.
		if (this._logoContainer) {
			map._controlCorners.bottomleft.appendChild(this._logoContainer);
		}
		if (this._attributionContainer) {
			map._controlCorners.bottomright.appendChild(this._attributionContainer);
		}

		//handle layer being added to a map for which there are no Google tiles at the given zoom
		google.maps.event.addListenerOnce(this._mutant, "idle", () => {
			if (!this._map) {
				return;
			}
			this._checkZoomLevels();
			this._mutantIsReady = true;
		});
	},

	_initMutant: function () {
		var map = new google.maps.Map(this._mutantContainer, {
			center: { lat: 0, lng: 0 },
			zoom: 0,
			tilt: 0,
			mapTypeId: this.options.type,
			disableDefaultUI: true,
			keyboardShortcuts: false,
			draggable: false,
			disableDoubleClickZoom: true,
			scrollwheel: false,
			styles: this.options.styles || [],
			backgroundColor: "transparent",
		});

		this._mutant = map;

		if (this._map) {
			this._mutantOnAdd();
		}
		// 🍂event spawned
		// Fired when the mutant has been created.
		this.fire("spawned", { mapObject: map });

		this._waitControls();
		this.once('controls_ready', this._setupAttribution);
	},

	// overridable method
	initApi: function (ready) {
		let checkCounter = 0,
			intervalId = null;

		intervalId = setInterval(() => {
			if (checkCounter >= 20) {
				clearInterval(intervalId);
				throw new Error("window.google not found after 10 seconds");
			}
			if (window.google && window.google.maps && window.google.maps.Map) {
				clearInterval(intervalId);
				ready();
			}
			++checkCounter;
		}, 500);
	},

	_attachObserver: function _attachObserver(node) {
		if (!this._observer) this._observer = new MutationObserver(this._onMutations.bind(this));

		// pass in the target node, as well as the observer options
		this._observer.observe(node, { childList: true, subtree: true });

		// if we are reusing an old _mutantContainer, we must manually detect
		// all existing tiles in it
		Array.prototype.forEach.call(node.querySelectorAll("img"), this._boundOnMutatedImage);
	},

	_waitControls: function () {
		const id = setInterval(() => {
			const layoutManager = this._mutant.__gm.layoutManager;
			if (!layoutManager) { return; }
			clearInterval(id);
			let positions;
			// iterate through obfuscated key names to find positions set (atm: layoutManager.o)
			Object.keys(layoutManager).forEach(function(key) {
				const el = layoutManager[key];
				if (el.get) {
					if (el.get(1) instanceof Node) {
						positions = el;
					}
				}
			});
			// 🍂event controls_ready
			// Fired when controls positions get available (passed in `positions` property).
			this.fire("controls_ready", { positions });
		}, 50);
	},

	_setupAttribution: function (ev) {
		if (!this._map) {
			return;
		}
		// https://developers.google.com/maps/documentation/javascript/reference/control#ControlPosition
		const pos = google.maps.ControlPosition;
		const ctr = this._attributionContainer = ev.positions.get(pos.BOTTOM_RIGHT);
		L.DomUtil.addClass(ctr, "leaflet-control leaflet-control-attribution");
		L.DomEvent.disableClickPropagation(ctr);
		ctr.style.height = "14px";
		this._map._controlCorners.bottomright.appendChild(ctr);

		this._logoContainer = ev.positions.get(pos.BOTTOM_LEFT);
		this._logoContainer.style.pointerEvents = "auto";
		this._map._controlCorners.bottomleft.appendChild(this._logoContainer);
	},

	_onMutations: function _onMutations(mutations) {
		for (var i = 0; i < mutations.length; ++i) {
			var mutation = mutations[i];
			for (var j = 0; j < mutation.addedNodes.length; ++j) {
				var node = mutation.addedNodes[j];

				if (node instanceof HTMLImageElement) {
					this._onMutatedImage(node);
				} else if (node instanceof HTMLElement) {
					Array.prototype.forEach.call(
						node.querySelectorAll("img"),
						this._boundOnMutatedImage
					);
				}
			}
		}
	},

	// Only images which 'src' attrib match this will be considered for moving around.
	// Looks like some kind of string-based protobuf, maybe??
	// Only the roads (and terrain, and vector-based stuff) match this pattern
	_roadRegexp: /!1i(\d+)!2i(\d+)!3i(\d+|VinaFnapurmBegrtn)!/,

	// On the other hand, raster imagery matches this other pattern
	_satRegexp: /x=(\d+)&y=(\d+)&z=(\d+|VinaFnapurmBegrtn)/,

	_onMutatedImage: function _onMutatedImage(imgNode) {
		let coords;
		let match = imgNode.src.match(this._roadRegexp);
		let sublayer = 0;

		if (match) {
			coords = {
				z: match[1],
				x: match[2],
				y: match[3],
			};
			if (this._imagesPerTile > 1) {
				imgNode.style.zIndex = 1;
				sublayer = 1;
			}
		} else {
			match = imgNode.src.match(this._satRegexp);
			if (match) {
				coords = {
					x: match[1],
					y: match[2],
					z: match[3],
				};
			}
			// imgNode.style.zIndex = 0;
			sublayer = 0;
		}

		if (coords) {
			var tileKey = this._tileCoordsToKey(coords);
			imgNode.style.position = "absolute";

			var key = tileKey + "/" + sublayer;
			// Cache img so it can also be used in subsequent tile requests
			this._lru.set(key, imgNode);

			if (key in this._tileCallbacks && this._tileCallbacks[key]) {
				// Use the tile for *all* pending callbacks. They'll be cloned anyway.
				this._tileCallbacks[key].forEach((callback) => callback(imgNode));
				delete this._tileCallbacks[key];
			}
		}
	},

	createTile: function (coords, done) {
		const key = this._tileCoordsToKey(coords),
			tileContainer = L.DomUtil.create("div");

		tileContainer.style.textAlign = "left";
		tileContainer.dataset.pending = this._imagesPerTile;
		done = done.bind(this, null, tileContainer);

		for (var i = 0; i < this._imagesPerTile; ++i) {
			const key2 = key + "/" + i,
				imgNode = this._lru.get(key2);
			if (imgNode) {
				tileContainer.appendChild(this._clone(imgNode));
				--tileContainer.dataset.pending;
			} else {
				this._tileCallbacks[key2] = this._tileCallbacks[key2] || [];
				this._tileCallbacks[key2].push(
					function (c /*, k2*/) {
						return function (imgNode) {
							c.appendChild(this._clone(imgNode));
							--c.dataset.pending;
							if (!parseInt(c.dataset.pending)) {
								done();
							}
						}.bind(this);
					}.bind(this)(tileContainer /*, key2*/)
				);
			}
		}

		if (!parseInt(tileContainer.dataset.pending)) {
			L.Util.requestAnimFrame(done);
		}
		return tileContainer;
	},

	_clone: function (imgNode) {
		const clonedImgNode = imgNode.cloneNode(true);
		clonedImgNode.style.visibility = "visible";
		return clonedImgNode;
	},

	_checkZoomLevels: function () {
		//setting the zoom level on the Google map may result in a different zoom level than the one requested
		//(it won't go beyond the level for which they have data).
		const zoomLevel = this._map.getZoom(),
			gMapZoomLevel = this._mutant.getZoom();

		if (!zoomLevel || !gMapZoomLevel) return;

		if (
			gMapZoomLevel !== zoomLevel || //zoom levels are out of sync, Google doesn't have data
			gMapZoomLevel > this.options.maxNativeZoom
		) {
			//at current location, Google does have data (contrary to maxNativeZoom)
			//Update maxNativeZoom
			this._setMaxNativeZoom(gMapZoomLevel);
		}
	},

	_setMaxNativeZoom: function (zoomLevel) {
		if (zoomLevel !== this.options.maxNativeZoom) {
			this.options.maxNativeZoom = zoomLevel;
			this._resetView();
		}
	},

	_update: function (center) {
		// zoom level check needs to happen before super's implementation (tile addition/creation)
		// otherwise tiles may be missed if maxNativeZoom is not yet correctly determined
		if (this._mutant) {
			center = center || this._map.getCenter();
			const _center = new google.maps.LatLng(center.lat, center.lng),
				zoom = Math.round(this._map.getZoom()),
				mutantZoom = this._mutant.getZoom();

			this._mutant.setCenter(_center);

			//ignore fractional zoom levels
			if (zoom !== mutantZoom) {
				this._mutant.setZoom(zoom);

				if (this._mutantIsReady) this._checkZoomLevels();
				//else zoom level check will be done later by 'idle' handler
			}
		}

		L.GridLayer.prototype._update.call(this, center);
	},

	// @method whenReady(fn: Function, context?: Object): this
	// Runs the given function `fn` when the mutant gets initialized, or immediately
	// if it's already initialized, optionally passing a function context.
	whenReady: function (callback, context) {
		if (this._mutant) {
			callback.call(context || this, { target: this });
		} else {
			this.on("spawned", callback, context);
		}
		return this;
	},
});

// 🍂factory gridLayer.googleMutant(options)
// Returns a new `GridLayer.GoogleMutant` given its options
L.gridLayer.googleMutant = function (options) {
	return new L.GridLayer.GoogleMutant(options);
};
